extensions/wikia/FilePage/FilePageController.class.php

```diff
@@ -73,6 +73,92 @@ public function fileUsage() {
 		wfProfileOut( __METHOD__ );
 	}
 
+	/**
+	 * Collects data about what articles the current file appears in, either
+	 * locally and globally.
+	 *
+	 * @requestParam string type (optional) Possible values are 'global' or 'local', defaulting to 'local' if not given.
+	 *                                      Determines whether this collects a local or global file usage data
+	 */
+	public function fileRedir() {
+		wfProfileIn( __METHOD__ );
+		global $wgMemc;
+
+		$page = $this->getContext();
+
+		//fallback to main page
+		$url = Title::newMainPage()->getFullURL();
+		//wiki needs read privileges
+		if ( !$page->getTitle()->userCan( 'read' ) ) {
+			$this->url = $url;
+			return;
+		}
+		$redirKey = wfMemcKey( 'redir', $page->getTitle()->getPrefixedText() );
+
+		$displayImg = $img = false;
+		Hooks::run( 'ImagePageFindFile', [ $page, &$img, &$displayImg ] );
+		if ( !$img ) { // not set by hook?
+			$img = wfFindFile( $page->getTitle() );
+			if ( !$img ) {
+				$img = wfLocalFile( $page->getTitle() );
+			}
+		}
+
+		if ( !$img || $img && !$img->fileExists ) {
+			$wgMemc->add( $redirKey, $url );
+			$this->url = $url;
+			return;
+		}
+
+		$urlMem = $wgMemc->get( $redirKey );
+		if ( $urlMem ) {
+			$this->url = $urlMem;
+			return;
+		}
+
+		$res = $this->fetchLinks( $img->getTitle()->getDBkey() );
+
+		foreach ( $res as $row ) {
+			$title = Title::newFromRow( $row );
+			if ( $title->isRedirect() ) {
+				continue;
+			}
+			if ( !$title->userCan( 'read' ) ) {
+				continue;
+			}
+			$url = $title->getFullURL();
+			break;
+		}
+		$wgMemc->add( $redirKey, $url );
```

Comment:
Shouldn't we check if `$url` is actually populated correctly? In worst case it will be the value from `Title::newMainPage()->getFullURL()` which can cause some issues.

Reply:
I've changed the logic a bit so everything works on $this->url. It must be populated at start with the mainpage url. The changes to image purges should take care of redirecting to the mainpage. The problem I can see is deploing this branch without PLATFORM-3932 @Grunny . After memcache stores a mainpage redirect it will stay there until it's purged.